File: python/analysis.py
import numpy as np
from numpy import dot
from numpy.random import randn
from numpy.linalg import inv
from math import sqrt

from observation import observe as H

RHO = 1

def enkf_assimilate(ensemble, obs_vec, Ro, sigma_o):
    # Mean ensemble vector
    ens_mean = np.mean(ensemble, axis=1)

    # Number of ensemble members
    N = ensemble.shape[1]

    # Dimension of observed vector
    num_obs = obs_vec.shape[0]

    # Table of perturbed observations. Each column contains the 
    # observation vector for this timestep with a bit of Gaussian-drawn
    # noise (different noise for each member)
    obs_table = np.empty((num_obs, N))
    for i in range(N):
        obs_table[:,i] = obs_vec + sigma_o * randn(num_obs)

    # 'Anomaly table'. Each column is that member's difference from the mean
    anom_table = ensemble - ens_mean[:,np.newaxis]

    # Ensemble covariance times transpose of observation matrix
    ens_cov_H_T = (RHO/(N-1)) * dot(anom_table, np.transpose(H(anom_table)))

    # Kalman gain
    gain = dot(ens_cov_H_T, inv(H(ens_cov_H_T) + Ro))

    # Combine previous ensemble and observations with Kalman gain to generate
    # next ensemble
    return ensemble + dot(gain, obs_table - H(ensemble))

def ensrf_assimilate(ensemble, obs_vec, Ro, sigma_o):
    # Mean ensemble vector
    ens_mean = np.mean(ensemble, axis=1)

    # Number of ensemble members
    m = ensemble.shape[1]

    # Observation dimension
    p = obs_vec.shape[0]

    # Forecast perturbation matrix
    Z_f = ensemble - ens_mean[:,np.newaxis]

    # Ensemble covariance times transpose of observation matrix
    ens_cov_H_T = dot(Z_f, np.transpose(H(Z_f))) / (m-1)

    # Kalman gain
    gain = dot(ens_cov_H_T, inv(H(ens_cov_H_T) + Ro))

    # Compute analysis mean
    ens_mean += dot(gain, obs_vec - H(ens_mean))

    V = H(Z_f)

    # Assimilate observations serially, as they are uncorrelated
    for i in range(p):
        v = V[i,:]
        v_dot = dot(v, v)

        D = v_dot + sigma_o[i]**2

        beta = 1.0/(D + sqrt(D*sigma_o[i]**2))

        Z_f = (1 - beta*v_dot) * Z_f


    return ens_mean[:,np.newaxis] + Z_f

File: fortran/analysis.f90
module analysis
    use params
    use utils

    implicit none

    contains
        pure function observe(state)
            real(dp), dimension(:, :), intent(in) :: state
            real(dp), dimension(obs_dim, size(state, 2)) :: observe

            observe = state(n_x+1:n_x+n_x*n_y,:)
        end function observe

        function assimilate(ensemble, obs_vec, obs_covar) result(analy)
            real(dp), dimension(state_dim, n_ens), intent(in) :: ensemble
            real(dp), dimension(obs_dim), intent(in) :: obs_vec
            real(dp), dimension(obs_dim, obs_dim), intent(in) :: obs_covar
            real(dp), dimension(state_dim, n_ens) :: analy, A
            real(dp), dimension(state_dim) :: ens_mean
            real(dp), dimension(obs_dim, n_ens) ::  obs_table
            real(dp), dimension(state_dim, obs_dim) :: gain
            real(dp), dimension(state_dim, obs_dim) :: ens_cov_h_t
            real(dp) :: y_std
            integer :: i, j

            ! Get y standard deviation (assume all y variances in y covariance matrix are the same)
            y_std = sqrt(obs_covar(1, 1))

            ! Mean ensemble vector
            ens_mean = (/ (sum(ensemble(i, :))/real(n_ens) , i = 1, state_dim) /)

            ! Table of perturbed observations
            do i = 1, n_ens
                do j = 1, obs_dim
                    obs_table(j, i) = obs_vec(j) + randn(0.0d0, y_std)
                end do
            end do

            ! 'Anomaly table'
            do i = 1, n_ens
                A(:, i) = ensemble(:, i) - ens_mean
            end do

            ! Ensemble covariance times transpose of observation matrix
            ens_cov_h_t = (rho/(n_ens-1)) * matmul(A, transpose(observe(A)))

            ! Kalman gain
            gain = matmul(ens_cov_h_t, inv(observe(ens_cov_h_t) + obs_covar))

            analy = ensemble + matmul(gain, obs_table - observe(ensemble))
        end function assimilate
end module analysis
